- Working with Files
- Workign with Libraries
- Installing the requests library with pip
- PRACTICE

# WORKING WITH FILES

- Context Managers

Python provides a built-in function for opening files.

- open()
- read()
- close()

By default, `open()` will open a file in read-only mode, however you can change this by passing a mode parameter. 

The list of optional modes is here:
|Character|Meaning|
|---|---|
|'r'|open for reading (default)|
|'w'|open for writing, truncating the file first|
|'x'|open for exclusive creation, failing if the file already exists|
|'a'|open for writing, appending to the end of the file if it exists|
|'b'|binary mode|
|'t'|text mode (default)|
|'+'|open a disk file for updating (reading and writing)|

```py
# Open a file for reading
>>> my_file = open("my_file.txt")

# Open a file for reading or writing
# This will replace any existing file
>>> my_file = open("my_file.txt", "w")

# Open a file for reading or writing
# This will append to the end of any existing file
>>> my_file = open("my_file.txt", "a")
```

But what happens if your program exits or crashes before you can close a file? 

Use a context manager!

## Context Managers

 Context Manager is like a wrapper around a block of code that depends on some resource. 

 It’s a safer way of handling resources than, say, using open() and then remembering to close() later (and hoping your program doesn’t crash in between).

 It’s similar to using try... finally.

 Context managers can contain code that auto-magically provisions a resource before your code runs, and cleans up afterward. 
 
 **Example** the open() function also works as a context manager, so opening a file looks like this:

 ```py
 >>> with open("my_file.text") as my_file:
...     contents = my_file.read()
```

 >You can perform all the functions that require access to the open file within the `with` scope, 

 >and once you exit the with scope, the context manager will `automatically close your file for you`, even if your code hits an exception inside the context manager.

 ### Working with Files
Let’s try working with a simple json file. 

We’ll import the json module to help us parse the data, and we’ll use a context manager to open the file.
```py
>>> import json
>>> with open("cities.json") as cities_file:
...     cities_data = json.load(cities_file)
...     print(cities_data)
...
[{'name': 'New York', 'pop': 8550405}, {'name': 'Los Angeles', 'pop': 3971883}, {'name': 'Chicago', 'pop': 2720546}, {'name': 'Houston', 'pop': 2296224}, {'name': 'Philadelphia', 'pop': 1567442}]
```

# WORKING WITH LIBRARIES

Working with external libraries in Python makes use of the `import` keyword.

#### different ways to import
- `import <module>` syntax

    Importing modules with the import keyword is usually the best method, because it preserves the module’s namespace. 

    ```py
    >>> import random
    >>> random.randint(0, 100)  # Here, namespace is preserved
    42
    ```

- `from <module> import <object>` syntax 
    
    to import a specific object (function, variable, subclass, etc.) from a module into your program’s namespace.

    ```py
    >>> from random import randint
    >>> randint(0, 100)  # namespace Not preserved
    64
    ```

    # Installing the requests library with pip

     3rd part library called requests - for retrieving data from web APIs. 

     To install the requests library, run this on your command line:

    ```(env) $ python3.7 -m pip install requests```

    This runs the pip module and asks it to find the requests library on PyPI.org (the Python Package Index) and install it in your local system, so that it becomes available for you to import. 

# PRACTICE

file_test_1.py
```py
    def main():
        pass

    if __name__ == "__main__":
        main()    
```

When this file_test.py is executed, as "python file_test.py",
it does northing.

Because you ran the file directly, the file’s __name__ variable is set to __main__, which triggers the if statement to run the main() function.

The `pass` keyword does nothing, it’s just there to keep the empty main() function from throwing a syntax error.

file_test_2.py
```py
import json

def main():
    cities_file = open("cities.json")
    cities_data = json.load(cities_file)
    print(cities_data)

if __name__ == "__main__":
    main()

# Output
python file_test_2.py

[{'name': 'New York', 'pop': 8550405}, {'name': 'Los Angeles', 'pop': 3971883}, {'name': 'Chicago', 'pop': 2720546}, {'name': 'Houston', 'pop': 2296224}, {'name': 'Philadelphia', 'pop': 1567442}]    
```
`enumerate() outputs a tuple of (index, entry)`, so we use index and entry variables to capture those.

file_test_3.py
```py
import json

def main():
    cities_file = open("cities.json")
    cities_data = json.load(cities_file)

    print("Largest cities in the US by population:")

    for index, entry in enumerate(cities_data):
        print(f"{index + 1}: {entry['name']} - {entry['pop']}")

if __name__ == "__main__":
    main()

# Output
python file_test_3.py

Largest cities in the US by population:
1: New York - 8550405
2: Los Angeles - 3971883
3: Chicago - 2720546
4: Houston - 2296224
5: Philadelphia - 1567442
```
#### Clean-up:
One more thing to clean up - using the `open()` keyword on its own is frowned upon, because it won’t automatically close any resources you might open. Even if you call the `close()` keyword yourself, there’s no guarantee your program won’t crash, leaving important resources dangling. It’s safer to open files inside a context using the `with` keyword. Once your code exits the scope of the context, your file is `automatically closed`.

file_test_4.py
```py
import json

def main():
    with open("cities.json") as cities_file:
        cities_data = json.load(cities_file)

        print("Largest cities in the US by population:")
        for index, entry in enumerate(cities_data):
            print(f"{index + 1}: {entry['name']} - {entry['pop']}")

    print("The file is now closed.")

if __name__ == "__main__":
    main()

# Output

python file_test_4.py

Largest cities in the US by population:
1: New York - 8550405
2: Los Angeles - 3971883
3: Chicago - 2720546
4: Houston - 2296224
5: Philadelphia - 1567442
The file is now closed.
```

if there is any error in the json format file, it will throw an error.

Let’s catch that JSONDecodeError, uisng try-catch.

file_test_5.py
```py
import json

def main():
    with open("cities.json") as cities_file:
        try:
            cities_data = json.load(cities_file)

            print("Largest cities in the US by population:")
            for index, entry in enumerate(cities_data):
                print(f"{index + 1}: {entry['name']} - {entry['pop']}")

        except json.decoder.JSONDecodeError as error:
            print("Sorry, there was an error decoding that json file:")
            print(f"\t {error}")

    print("The file is now closed.")

if __name__ == "__main__":
    main()

# Output

python file_execise.py
Sorry, there was an error decoding that json file:
	 Expecting ',' delimiter: line 21 column 1 (char 234)
The file is now closed.
```